wp-includes/class-wp-offline-page-excluder.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * WP_Offline_Page_Excluder class.
+ *
+ * @package PWA
+ */
+
+/**
+ * Class handles excluding the Offline Page's from the backend and frontend.
+ */
+class WP_Offline_Page_Excluder {
+
+	/**
+	 * Instance of the Offline Page Manager.
+	 *
+	 * @var WP_Offline_Page
+	 */
+	protected $manager;
+
+	/**
+	 * WP_Offline_Page_Filter constructor.
+	 *
+	 * @param WP_Offline_Page $manager Instance of the manager.
+	 */
+	public function __construct( $manager ) {
+		$this->manager = $manager;
+	}
+
+	/**
+	 * Initializes the instance.
+	 */
+	public function init() {
+		add_filter( 'wp_dropdown_pages', array( $this, 'exclude_from_page_dropdown' ), 10, 2 );
+		add_action( 'parse_query', array( $this, 'exclude_from_query' ) );
+	}
+
+	/**
+	 * Filters the HTML output of a list of pages as a drop down.
+	 *
+	 * @param string $html HTML output for drop down list of pages.
+	 * @param array  $args The parsed arguments array.
+	 * @return string Filtered content for wp_dropdown_pages.
+	 */
+	public function exclude_from_page_dropdown( $html, $args ) {
+		// Bail out if this is the offline page dropdown.
+		if ( WP_Offline_Page::OPTION_NAME === $args['name'] ) {
+			return $html;
+		}
+
+		return preg_replace( '/<option .*? value="' . $this->manager->get_offline_page_id() . '">.*?<\/option>/', '', $html );
+	}
+
+	/**
+	 * Exclude the offline page from the query when doing a search on the frontend or on the backend Menus page.
+	 *
+	 * @param WP_Query $query The WP_Query instance.
+	 */
+	public function exclude_from_query( WP_Query $query ) {
+		if ( $this->is_offline_page_query( $query ) ) {
+			$query->is_404      = true;
+			$query->is_page     = false;
+			$query->is_singular = false;
+			$query->set( 'page_id', 0 );
```

Comment:
Are these being done in anticipation of there being an `is_offline`? 

I don't think that we should necessarily make these changes here. The offline page is a page, so `is_page()` should return `true`. But what we could do here is `$query->is_offline = $this->is_offline_page_query( $query )` to dynamically define a new member variable that can be looked at similarly to the others. We can then introduce a new global function:

```php
function is_offline() {
	global $wp_query;
	if ( ! isset( $wp_query ) ) {
		_doing_it_wrong( __FUNCTION__, __( 'Conditional query tags do not work before the query is run. Before then, they always return false.' ), '3.1.0' );
		return false;
	}
	return ! empty( $wp_query->is_offline );
}
```

And this `is_offline()` function can be used in the template to conditionally show content for offline responses.

Reply:
>Are these being done in anticipation of there being an is_offline?

@westonruter No, it's meant to prevent the offline page from being displayed on the frontend. 
 It's forcing a 404.  It was added to provide the functionality [discussed here in Issue #23](https://github.com/xwp/pwa-wp/issues/23#issuecomment-404395715).

>I don't think that we should necessarily make these changes here.

Okay.  Let's think about this.  Maybe we don't care if the default offline page is requestable or viewable on the frontend.

- We're already excluding it from the query via `post__not_in`.
- We exclude it from the menus.

What's the harm of viewing it on the frontend?  If no or little, then there's no need to set a 404 here.

>And this is_offline() function can be used in the template to conditionally show content for offline responses.

Agreed.  I was thinking about adding that functionality in a separate PR.  We'll want pages/content be marked as `offline` and then set a var to note it.

Reply:
I think there is no need to prevent the offline page from being accessible if someone knows the URL. If the user can access it directly by the URL, that is fine. In fact, that should be actually desirable because it would allow for someone to easily preview their offline page and make changes. The main thing is just that we don't expose the offline page in menus, page lists, searches, etc. In case someone links to the offline page then we could add `wp_no_robots()` to run in `wp_head` for that specific page.

Reply:
Bam, done in [commit ab4cbc6](https://github.com/xwp/pwa-wp/pull/28/commits/ab4cbc66a3d61e809da99451d2e1a63f17cfa66a).  The _default_ offline page will now render on the frontend.  That makes it easier.  

> In case someone links to the offline page then we could add wp_no_robots() to run in wp_head for that specific page.

Included in this [commit 9669026](https://github.com/xwp/pwa-wp/pull/28/commits/96690264eed185fd748ce763d3049aaa2e74c655).